packages/pdf-viewer/utils/renderQueue.ts

```diff
@@ -0,0 +1,39 @@
+import { RenderRequest, RenderResult } from '../types';
+
+interface Task {
+	resolve: (result: RenderResult)=> void;
+	reject: (error: any)=> void;
+	params: RenderRequest;
+}
+
+export default class RenderQueue {
+	private tasks: Task[];
+	private lock = false;
+	private executeTaskImpl: (params: RenderRequest)=> Promise<RenderResult>;
+
+	public constructor(executeTaskImpl: (params: RenderRequest)=> Promise<RenderResult>) {
+		this.tasks = [];
+		this.executeTaskImpl = executeTaskImpl;
+	}
+
+	public addTask = (task: Task) => {
+		if (this.lock) {
+			// Some task is already in progress, so add this task to the queue
+			this.tasks.push(task);
+		} else {
+			this.lock = true;
+
+			const next = () => {
+				this.lock = false;
+				if (this.tasks.length > 0) {
+					// Time to execute the next task in the queue
+					const nextTask = this.tasks.shift();
+					this.addTask(nextTask);
+				}
+			};
+
+			// Render the page
+			this.executeTaskImpl(task.params).then(task.resolve).catch(task.reject).finally(next);
+		}
+	};
+}
```

Comment:
Since this is used in an async context, there are race conditions in there and you'll end up with dead locks, or task not being rendered. Not sure what to suggest, you need to reduce the complexity of all this, maybe by getting rid of this render queue, and rendering in a sequential way.

It's not clear to me why you need a queue anyway, can't you just render what's current being displayed?

Reply:
The renderer queue's job is to make sure rendering happens sequentially. 

The issue is that since in full screen viewer there are thumbnails as well as the main view and all of them fire the rendering simultaneously. For example page 1 in main view and page 1 in thumbnail view fires simultaneously (before waiting for the other to finish) and I believe this is an issue with the pdfjs library that when such a scenario happens (rendering same page) the later one fails without throwing any errors. I had to experiment a lot to figure out the issue. 

Infact I saw that the sample pdfjs viewer also uses a rendering queue distinguishing between thumbnails and main view:
https://github.com/mozilla/pdf.js/blob/master/web/pdf_rendering_queue.js

> Since this is used in an async context, there are race conditions in there and you'll end up with dead locks,

I'm not sure I follow, the queue exists to eliminate the race condition, can you elaborate a bit? Or any articles to follow.

Reply:
The problem is that pdfjs is going to trigger rendering tasks in an async way and your code currently has a race condition. For example if immediately after the check on line 28 a new task is pushed, that task will never be rendered.

Getting concurrent code like this right is hard and a simple lock=true/false is not going to do it. Ideally we wouldn't have this code at all here but if it's required then you might want to look at how to write proper concurrent code, in particular look at "Producer–consumer problem" and synchronisation primitives such as semaphores.

In Joplin we use the async-mutex package in various places in the code, perhaps you can use that too.

Reply:
Sounds interesting, will study the matter and try to come up with a better implementation for this. Thanks!